lib/semantics/check-omp-structure.cc

```diff
@@ -114,18 +113,57 @@ void OmpStructureChecker::Enter(const parser::OmpLoopDirective::Do &) {
   // reserve for nesting check
   SetContextDirectiveEnum(OmpDirective::DO);
 
-  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::LASTPRIVATE,
-      OmpClause::LINEAR, OmpClause::REDUCTION, OmpClause::SCHEDULE,
-      OmpClause::COLLAPSE, OmpClause::ORDERED};
+  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::FIRSTPRIVATE,
+      OmpClause::LASTPRIVATE, OmpClause::LINEAR, OmpClause::REDUCTION};
   SetContextAllowed(allowed);
 
   OmpClauseSet allowedOnce{
       OmpClause::SCHEDULE, OmpClause::COLLAPSE, OmpClause::ORDERED};
   SetContextAllowedOnce(allowedOnce);
 }
 
+void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
+  // 2.7 Loop Construct Restriction
+  if (GetContext().directive == OmpDirective::DO) {
+    if (SeenClause(OmpClause::SCHEDULE)) {
+      // only one schedule clause is allowed
+      auto it{GetContext().clauseInfo.find(OmpClause::SCHEDULE)};
+      auto *clause{it->second};
+      const auto &schedClause{std::get<parser::OmpScheduleClause>(clause->u)};
+      if (ScheduleModifierHasType(schedClause,
+              parser::OmpScheduleModifierType::ModType::Nonmonotonic)) {
+        if (SeenClause(OmpClause::ORDERED)) {
+          context_.Say(clause->source,
+              "The NONMONOTONIC modifier cannot be specified "
+              "if an ORDERED clause is specified"_err_en_US);
+        }
+        if (ScheduleModifierHasType(schedClause,
+                parser::OmpScheduleModifierType::ModType::Monotonic)) {
+          context_.Say(clause->source,
+              "The MONOTONIC and NONMONOTONIC modifiers "
+              "cannot be both specified"_err_en_US);
+        }
+      }
+    }
+
+    if (SeenClause(OmpClause::ORDERED) && SeenClause(OmpClause::LINEAR)) {
+      // only one ordered clause is allowed
+      auto it{GetContext().clauseInfo.find(OmpClause::ORDERED)};
+      auto *clause{it->second};
+      const auto &orderedClause{
+          std::get<parser::OmpClause::Ordered>(clause->u)};
+      if (orderedClause.v.has_value()) {
+        context_.Say(clause->source,
+            "A LINEAR clause or an ORDERED clause with a parameter "
+            "can be specified on a loop directive but not both"_err_en_US);
```

Comment:
How about:
"A loop directive may not have both a LINEAR clause and an ORDERED clause with a parameter"

Reply:
Addressed in commit [1b57193](https://github.com/flang-compiler/f18/pull/536/commits/1b57193d2377c5121f16075950b64ea77fa72e2e).